Markdown Checkbox Functionality

AdritRao opened the pull request:
# Markdown Checkbox Functionality

## :recycle: Current situation & Problem
Checkboxes are currently not supported in markdown consent forms for SpeziOnboarding. Adding support for checkbox syntax in MD files can allow for yes/no consent questions in addition to the name and signature.

## :gear: Release Notes 

- Added functionality for checkbox syntax to be extracted from MD (`extractMarkdownCB()`). Checkbox questions can be put within square brackets (e.g. [Do you consent to data collection?]). 
- `extractMarkdownCB()` also returns a cleaned MD with checkbox syntax removed for PDF export
- Added `checkboxesView` to create checkbox buttons based on the extracted items from MD
- Used `ImageRenderer` to capture snapshot of `checkboxesView` when user taps 'I Consent'
- Updated document export to use cleaned MD and added image of checkbox snapshot under the MD text (when available)
- Updated `OnboardingConsentMarkdownRenderingView.swift` to allow for QuickLook PDF view

Below is an image of the UI and PDF export produced from the following MD data:
```
Data(
    """
    This is a *markdown* **example** of a consent document. 
    [May we contact you for future studies?] 
    [Can we store your step count for research purposes?]
    """.utf8
)
```

<img src="https://github.com/user-attachments/assets/2168cb24-6178-4e32-9ff0-9f97a38a74c8" alt="Simulator Screenshot - iPhone 15 Pro - 2024-08-02 at 01 21 58" width="200">

<img src="https://github.com/user-attachments/assets/5ce28a7e-394b-4852-bf55-8a000f1639ff" alt="Simulator Screenshot - iPhone 15 Pro - 2024-08-02 at 01 28 54" width="200">

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

PSchmiedmayer reviewed (changes requested):
@AdritRao Thank you for taking a look at the issue and addressing a first element here!

I have a few first things that could help as you move forward:
1. I would suggest using the GitHub task list syntax to extract checkboxes from the markdown: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#task-lists
2. I would suggest that we allow these checkboxes to be translated to either Toggles or a Menu. I would probably prefer a SwiftUI Menu as it can also detail that no selection has been made yet, which might be important if a user needs to either agree or disagree to something.
3. We should embed these at the places where they occur in the Markdown. This requires us to basically parse the MD fileline by line, then investigate if one line contains a special syntax we need to interpret (see above or below for possibilities), and then transform this into a specialized element. I would suggest that we create a special data structure for this instead of doing this in a few. This would also make it easier to unit test this:
    a. The data structure contains a list of elements that are enums. These can be Markdown text, an agree/disagree selection with a description, or the signature field (keeping it simple for these three for now). 
    b. You can parse an MD string into this format so it is transformed in our internal representation.
    c. This is then used in the Consent Document to basically render this in a ForEach loop that just renders each section one after each other in our VStack/ScrollView.
    d. On export, this data is then collected and transferred into an easily parsable representation as noted in #51.

In addition, we should take a look at the GitHub MD syntax on how we might represent additional fields like an initial signature field. I can imagine we can use: 
```md
> [!INITIALS]
> Description for the initials field.
```
Syntax or something similar to how GitHub displays alerts boxes: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts.

I can also see that we only rely on this syntax and also use that for selections:
```md
> [!AGREE]
> Description of the Agree/Disagree field
```
and 
```md
> [!SIGNATURE]
```
to even inject the signature field at the bottom of the consent document.

Let me know what you think; thank you for taking a deep-dive into all of this!

AdritRao commented:
Thank you for reviewing the PR and for providing these next steps @PSchmiedmayer! I will start working on migrating to the GitHub syntax and using SwiftUI menus. I will also add the functionality for line-by-line parsing to make sure that elements are created and displayed based on the order they appear in the Markdown file. After this, I will add the additional elements such as initials, etc.

philippzagar commented:
Hey @AdritRao, great PR so for! 🚀 Just wanted to follow up on this PR and see what the current status is?
Let me know if you need any input from me (besides the already extensive feedback from @PSchmiedmayer).

AdritRao commented:
Hey @philippzagar, thanks for bringing this up! I am aiming to get to this and wrap it up some time in the next two weeks. Will keep you posted!

AdritRao commented:
Hi @lukaskollmer,

I have had a chance to address the main PR comments.

This is what I have addressed based on @PSchmiedmayer's PR review:
- Adapted the markdown extraction code to the GitHub syntax for checkboxes (allowing for custom options)
- Using SwiftUI menu to allow for these custom options in the UI
- Doing line-by-line parsing of the markdown so that elements show up in the correct order. Created a enum ```MarkdownElement``` and an array ``` allElements``` of these enums that are populated in the extract function. The UI is created using a ForEach loop to render sequentially through each ```MarkdownElement```.

I have added to-do comments in the location of the code in ```ConsentDocument``` for the main additional features that need to be added.

These are the main features that need to be added:
- The validation component to ensure that all required fields are selected
- Rendering the PDF based on the completed consent form
- The options for additional MD fields as @PSchmiedmayer mentioned above (e.g. initials, agree etc.).

Currently, there are merge conflicts that I have not addressed. Please feel free to address them based on the changes that have been made to the codebase in the past few months.

Thank you!
